### PR TITLE
Implemented Offline Repair in the frontend for delete operation.

### PR DESCRIPTION
**Pull request by JingQianCloud**
If delete is partially failed and offline repair is enabled, insert one RepairRecord to the db and return success status for the delete.

Will create metrics in a separate PR.

### CODE REVIEW

File: ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java [hunk 1]
```diff
@@ -135,6 +135,12 @@ public class RouterConfig {
   public static final String ROUTER_REPAIR_WITH_REPLICATE_BLOB_ON_DELETE_ENABLED =
       "router.repair.with.replicate.blob.on.delete.enabled";
   public static final String ROUTER_REPAIR_WITH_REPLICATE_BLOB_ENABLED = "router.repair.with.replicate.blob.enabled";
+  // offline repair partially failed ttl update
+  public static final String ROUTER_TTLUPDATE_OFFLINE_REPAIR_ENABLED = "router.ttlupdate.offline.repair.enabled";
```
Comment (SophieGuo410) [inline ambry-api/src/main/java/com/github/ambry/config/RouterConfig.java:139]: This config is not used, are you planning to use in follow up pr?

File: ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java [hunk 2]
```diff
@@ -93,11 +99,11 @@ public class NonBlockingRouter implements Router {
    * @throws IOException if the OperationController could not be successfully created.
    * @throws ReflectiveOperationException if the OperationController could not be successfully created.
    */
-  public NonBlockingRouter(RouterConfig routerConfig, NonBlockingRouterMetrics routerMetrics,
-      NetworkClientFactory networkClientFactory, NotificationSystem notificationSystem, ClusterMap clusterMap,
-      KeyManagementService kms, CryptoService cryptoService, CryptoJobHandler cryptoJobHandler,
-      AccountService accountService, Time time, String defaultPartitionClass, AmbryCache blobMetadataCache)
-      throws IOException, ReflectiveOperationException {
+  public NonBlockingRouter(RouterConfig routerConfig, VerifiableProperties vProps,
```
[the remark this answers is not in the document]
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/NonBlockingRouter.java]: Fixed. Replaced with RepairRequestsDbFactory. Thanks,

File: ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java [hunk 3]
```diff
@@ -372,10 +380,32 @@ private boolean shouldReplicateBlobAndRetry() {
...
         && getPrecedenceLevel(errorCode) >= getPrecedenceLevel(RouterErrorCode.AmbryUnavailable)
         && (serviceId == null || !serviceId.startsWith(BackgroundDeleteRequest.SERVICE_ID_PREFIX)));
+    // @formatter:on
+  }
+
+  /**
+   * Check if we want to save the partially failed request to the secondary storage and fix it offline.
+   * @return true if choose to fix it offline.
+   */
+  private boolean shouldRunOfflineRepair() {
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-router/src/main/java/com/github/ambry/router/DeleteOperation.java:395]: hmmm, I think that's fine.

**Review by SophieGuo410 — APPROVED** [review]

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

**Pull request merged**